File: core/scanner/SteamScanner.cpp
#include "core/scanner/SteamScanner.h"

#include <algorithm>
#include <cctype>
#include <cstdlib>
#include <fstream>
#include <set>
#include <sstream>
#include <windows.h>

namespace gsm::core {
namespace {

std::string readTextFile(const std::string& path)
{
    std::ifstream file(path, std::ios::binary);
    if (!file) {
        return {};
    }

    std::ostringstream buffer;
    buffer << file.rdbuf();
    return buffer.str();
}

std::string registryString(HKEY root, const char* subKey, const char* valueName)
{
    char buffer[4096]{};
    DWORD bufferSize = sizeof(buffer);
    const LONG result = RegGetValueA(
        root,
        subKey,
        valueName,
        RRF_RT_REG_SZ,
        nullptr,
        buffer,
        &bufferSize);

    if (result != ERROR_SUCCESS || bufferSize == 0) {
        return {};
    }

    return std::string(buffer);
}

std::optional<std::string> environmentVariable(const char* name)
{
    const char* value = std::getenv(name);
    if (value == nullptr || value[0] == '\0') {
        return std::nullopt;
    }
    return std::string(value);
}

bool isNumeric(const std::string& value)
{
    return !value.empty() && std::all_of(value.begin(), value.end(), [](unsigned char character) {
        return std::isdigit(character) != 0;
    });
}

std::string toLower(std::string value)
{
    std::transform(value.begin(), value.end(), value.begin(), [](unsigned char character) {
        return static_cast<char>(std::tolower(character));
    });
    return value;
}

std::string unescapeVdfString(const std::string& value)
{
    std::string result;
    result.reserve(value.size());

    for (std::size_t index = 0; index < value.size(); ++index) {
        const char character = value[index];
        if (character != '\\' || index + 1 >= value.size()) {
            result += character;
            continue;
        }

        const char next = value[++index];
        switch (next) {
        case '\\':
            result += '\\';
            break;
        case '"':
            result += '"';
            break;
        case 'n':
            result += '\n';
            break;
        case 't':
            result += '\t';
            break;
        default:
            result += next;
            break;
        }
    }

    return result;
}

std::vector<std::string> tokenizeVdf(const std::string& text)
{
    std::vector<std::string> tokens;

    for (std::size_t index = 0; index < text.size();) {
        const char character = text[index];

        if (std::isspace(static_cast<unsigned char>(character)) != 0) {
            ++index;
            continue;
        }

        if (character == '/' && index + 1 < text.size() && text[index + 1] == '/') {
            while (index < text.size() && text[index] != '\n') {
                ++index;
            }
            continue;
        }

        if (character == '{' || character == '}') {
            tokens.emplace_back(1, character);
            ++index;
            continue;
        }

        if (character == '"') {
            ++index;
            std::string value;
            while (index < text.size()) {
                const char current = text[index++];
                if (current == '"') {
                    break;
                }
                if (current == '\\' && index < text.size()) {
                    value += current;
                    value += text[index++];
                    continue;
                }
                value += current;
            }
            tokens.push_back(unescapeVdfString(value));
            continue;
        }

        std::string bareToken;
        while (index < text.size()) {
            const char current = text[index];
            if (std::isspace(static_cast<unsigned char>(current)) != 0 || current == '{' || current == '}') {
                break;
            }
            bareToken += current;
            ++index;
        }
        tokens.push_back(bareToken);
    }

    return tokens;
}

void addUniquePath(std::vector<gsm::system::Path>& paths, const gsm::system::Path& path)
{
    const gsm::system::Path normalized = gsm::system::normalizePath(path);
    if (normalized.empty()) {
        return;
    }

    const std::string normalizedKey = toLower(normalized);
    const auto found = std::find_if(paths.begin(), paths.end(), [&normalizedKey](const gsm::system::Path& existing) {
        return toLower(gsm::system::normalizePath(existing)) == normalizedKey;
    });

    if (found == paths.end()) {
        paths.push_back(normalized);
    }
}

void addUniqueGame(std::vector<GameEntry>& games, const GameEntry& game)
{
    const std::string key = toLower(game.sourceId + "|" + gsm::system::normalizePath(game.installPath));
    const auto found = std::find_if(games.begin(), games.end(), [&key](const GameEntry& existing) {
        return toLower(existing.sourceId + "|" + gsm::system::normalizePath(existing.installPath)) == key;
    });

    if (found == games.end()) {
        games.push_back(game);
    }
}

} // namespace

std::string toString(GameSource source)
{
    switch (source) {
    case GameSource::Manual:
        return "manual";
    case GameSource::Steam:
        return "steam";
    case GameSource::Epic:
        return "epic";
    }

    return "unknown";
}

std::vector<GameEntry> SteamScanner::scanInstalledGames() const
{
    std::vector<GameEntry> games;
    std::vector<gsm::system::Path> libraries;

    for (const gsm::system::Path& steamRoot : discoverSteamRoots()) {
        for (const gsm::system::Path& library : discoverLibraryFolders(steamRoot)) {
            addUniquePath(libraries, library);
        }
    }

    for (const gsm::system::Path& library : libraries) {
        std::vector<GameEntry> libraryGames = scanLibrary(library);
        for (const GameEntry& game : libraryGames) {
            addUniqueGame(games, game);
        }
    }

    std::sort(games.begin(), games.end(), [](const GameEntry& left, const GameEntry& right) {
        return left.name < right.name;
    });

    return games;
}

std::vector<gsm::system::Path> SteamScanner::discoverSteamRoots() const
{
    std::vector<gsm::system::Path> roots;

    addUniquePath(roots, registryString(HKEY_CURRENT_USER, "Software\\Valve\\Steam", "SteamPath"));
    addUniquePath(roots, registryString(HKEY_LOCAL_MACHINE, "SOFTWARE\\WOW6432Node\\Valve\\Steam", "InstallPath"));
    addUniquePath(roots, registryString(HKEY_LOCAL_MACHINE, "SOFTWARE\\Valve\\Steam", "InstallPath"));

    if (const auto programFilesX86 = environmentVariable("ProgramFiles(x86)")) {
        addUniquePath(roots, gsm::system::joinPath(*programFilesX86, "Steam"));
    }

    if (const auto programFiles = environmentVariable("ProgramFiles")) {
        addUniquePath(roots, gsm::system::joinPath(*programFiles, "Steam"));
    }

    roots.erase(std::remove_if(roots.begin(), roots.end(), [](const gsm::system::Path& path) {
        return !gsm::system::directoryExists(path);
    }), roots.end());

    return roots;
}

std::vector<gsm::system::Path> SteamScanner::discoverLibraryFolders(const gsm::system::Path& steamRoot) const
{
    const gsm::system::Path libraryFoldersPath =
        gsm::system::joinPath(gsm::system::joinPath(steamRoot, "steamapps"), "libraryfolders.vdf");

    const std::string text = readTextFile(libraryFoldersPath);
    if (text.empty()) {
        return {};
    }

    return parseLibraryFoldersVdf(steamRoot, text);
}

std::vector<GameEntry> SteamScanner::scanLibrary(const gsm::system::Path& libraryPath) const
{
    const gsm::system::Path steamAppsPath = gsm::system::joinPath(libraryPath, "steamapps");
    std::vector<GameEntry> games;

    for (const gsm::system::Path& manifestPath : gsm::system::listFilesWithPrefixSuffix(steamAppsPath, "appmanifest_", ".acf")) {
        const std::string text = readTextFile(manifestPath);
        const std::optional<GameEntry> game = parseAppManifest(libraryPath, text);
        if (game.has_value()) {
            games.push_back(*game);
        }
    }

    return games;
}

std::vector<gsm::system::Path> SteamScanner::parseLibraryFoldersVdf(
    const gsm::system::Path& steamRoot,
    const std::string& text)
{
    std::vector<gsm::system::Path> libraries;
    addUniquePath(libraries, steamRoot);

    const std::vector<std::string> tokens = tokenizeVdf(text);

    for (std::size_t index = 0; index + 1 < tokens.size(); ++index) {
        const std::string& token = tokens[index];
        const std::string& next = tokens[index + 1];

        if (token == "path" && next != "{" && next != "}") {
            addUniquePath(libraries, next);
            continue;
        }

        if (isNumeric(token) && next != "{" && next != "}") {
            addUniquePath(libraries, next);
        }
    }

    return libraries;
}

std::optional<GameEntry> SteamScanner::parseAppManifest(
    const gsm::system::Path& libraryPath,
    const std::string& text)
{
    const std::vector<std::string> tokens = tokenizeVdf(text);

    std::string appId;
    std::string name;
    std::string installDir;

    for (std::size_t index = 0; index + 1 < tokens.size(); ++index) {
        const std::string& key = tokens[index];
        const std::string& value = tokens[index + 1];

        if (value == "{" || value == "}") {
            continue;
        }

        if (key == "appid") {
            appId = value;
        } else if (key == "name") {
            name = value;
        } else if (key == "installdir") {
            installDir = value;
        }
    }

    if (appId.empty() || name.empty() || installDir.empty()) {
        return std::nullopt;
    }

    GameEntry entry;
    entry.source = GameSource::Steam;
    entry.sourceId = appId;
    entry.name = name;
    entry.libraryPath = gsm::system::normalizePath(libraryPath);
    entry.installPath = gsm::system::joinPath(gsm::system::joinPath(entry.libraryPath, "steamapps\\common"), installDir);
    entry.installed = true;
    return entry;
}

} // namespace gsm::core

File: game_storage_manager/core/scanner/steam_scanner.py
from __future__ import annotations

import os
import winreg

from game_storage_manager.core.scanner.game_entry import GameEntry, GameSource
from game_storage_manager.system.filesystem.filesystem import (
    directory_exists,
    join_path,
    list_files_with_prefix_suffix,
    normalize_path,
)


def _read_text_file(path: str) -> str:
    try:
        with open(path, "r", encoding="utf-8", errors="replace") as f:
            return f.read()
    except OSError:
        return ""


def _registry_string(root: int, sub_key: str, value_name: str) -> str:
    try:
        key = winreg.OpenKey(root, sub_key)
        value, _ = winreg.QueryValueEx(key, value_name)
        winreg.CloseKey(key)
        return str(value) if value else ""
    except OSError:
        return ""


def _environment_variable(name: str) -> str | None:
    value = os.environ.get(name)
    return value if value else None


def _is_numeric(value: str) -> bool:
    return bool(value) and value.isdigit()


def _unescape_vdf_string(value: str) -> str:
    result = []
    i = 0
    while i < len(value):
        ch = value[i]
        if ch != "\\" or i + 1 >= len(value):
            result.append(ch)
            i += 1
            continue
        i += 1
        nxt = value[i]
        if nxt == "\\":
            result.append("\\")
        elif nxt == '"':
            result.append('"')
        elif nxt == "n":
            result.append("\n")
        elif nxt == "t":
            result.append("\t")
        else:
            result.append(nxt)
        i += 1
    return "".join(result)


def _tokenize_vdf(text: str) -> list[str]:
    tokens: list[str] = []
    i = 0
    while i < len(text):
        ch = text[i]

        if ch.isspace():
            i += 1
            continue

        if ch == "/" and i + 1 < len(text) and text[i + 1] == "/":
            while i < len(text) and text[i] != "\n":
                i += 1
            continue

        if ch in ("{", "}"):
            tokens.append(ch)
            i += 1
            continue

        if ch == '"':
            i += 1
            value_chars: list[str] = []
            while i < len(text):
                current = text[i]
                i += 1
                if current == '"':
                    break
                if current == "\\" and i < len(text):
                    value_chars.append(current)
                    value_chars.append(text[i])
                    i += 1
                    continue
                value_chars.append(current)
            tokens.append(_unescape_vdf_string("".join(value_chars)))
            continue

        bare: list[str] = []
        while i < len(text):
            current = text[i]
            if current.isspace() or current in ("{", "}"):
                break
            bare.append(current)
            i += 1
        tokens.append("".join(bare))

    return tokens


def _add_unique_path(paths: list[str], path: str) -> None:
    normalized = normalize_path(path)
    if not normalized:
        return
    key = normalized.lower()
    for existing in paths:
        if normalize_path(existing).lower() == key:
            return
    paths.append(normalized)


def _add_unique_game(games: list[GameEntry], game: GameEntry) -> None:
    key = (game.source_id + "|" + normalize_path(game.install_path)).lower()
    for existing in games:
        if (existing.source_id + "|" + normalize_path(existing.install_path)).lower() == key:
            return
    games.append(game)


class SteamScanner:
    def scan_installed_games(self) -> list[GameEntry]:
        libraries: list[str] = []
        for steam_root in self.discover_steam_roots():
            for library in self.discover_library_folders(steam_root):
                _add_unique_path(libraries, library)

        games: list[GameEntry] = []
        for library in libraries:
            for game in self.scan_library(library):
                _add_unique_game(games, game)

        games.sort(key=lambda g: g.name)
        return games

    def discover_steam_roots(self) -> list[str]:
        roots: list[str] = []

        _add_unique_path(
            roots, _registry_string(winreg.HKEY_CURRENT_USER, "Software\\Valve\\Steam", "SteamPath"))
        _add_unique_path(
            roots, _registry_string(
                winreg.HKEY_LOCAL_MACHINE, "SOFTWARE\\WOW6432Node\\Valve\\Steam", "InstallPath"))
        _add_unique_path(
            roots, _registry_string(winreg.HKEY_LOCAL_MACHINE, "SOFTWARE\\Valve\\Steam", "InstallPath"))

        pf_x86 = _environment_variable("ProgramFiles(x86)")
        if pf_x86:
            _add_unique_path(roots, join_path(pf_x86, "Steam"))

        pf = _environment_variable("ProgramFiles")
        if pf:
            _add_unique_path(roots, join_path(pf, "Steam"))

        roots = [r for r in roots if directory_exists(r)]
        return roots

    def discover_library_folders(self, steam_root: str) -> list[str]:
        vdf_path = join_path(join_path(steam_root, "steamapps"), "libraryfolders.vdf")
        text = _read_text_file(vdf_path)
        if not text:
            return []
        return self.parse_library_folders_vdf(steam_root, text)

    def scan_library(self, library_path: str) -> list[GameEntry]:
        steam_apps_path = join_path(library_path, "steamapps")
        games: list[GameEntry] = []
        for manifest_path in list_files_with_prefix_suffix(steam_apps_path, "appmanifest_", ".acf"):
            text = _read_text_file(manifest_path)
            game = self.parse_app_manifest(library_path, text)
            if game is not None:
                games.append(game)
        return games

    @staticmethod
    def parse_library_folders_vdf(steam_root: str, text: str) -> list[str]:
        libraries: list[str] = []
        _add_unique_path(libraries, steam_root)

        tokens = _tokenize_vdf(text)
        i = 0
        while i + 1 < len(tokens):
            token = tokens[i]
            nxt = tokens[i + 1]

            if token == "path" and nxt not in ("{", "}"):
                _add_unique_path(libraries, nxt)
                i += 1
                continue

            if _is_numeric(token) and nxt not in ("{", "}"):
                _add_unique_path(libraries, nxt)

            i += 1

        return libraries

    @staticmethod
    def parse_app_manifest(library_path: str, text: str) -> GameEntry | None:
        tokens = _tokenize_vdf(text)

        app_id = ""
        name = ""
        install_dir = ""

        i = 0
        while i + 1 < len(tokens):
            key = tokens[i]
            value = tokens[i + 1]

            if value in ("{", "}"):
                i += 1
                continue

            if key == "appid":
                app_id = value
            elif key == "name":
                name = value
            elif key == "installdir":
                install_dir = value

            i += 1

        if not app_id or not name or not install_dir:
            return None

        entry = GameEntry()
        entry.source = GameSource.Steam
        entry.source_id = app_id
        entry.name = name
        entry.library_path = normalize_path(library_path)
        entry.install_path = join_path(join_path(entry.library_path, "steamapps\\common"), install_dir)
        entry.installed = True
        return entry
